confection/util.py

```diff
@@ -90,6 +91,14 @@ def pop(self, key, default=None):
     def update(self, other):
         raise NotImplementedError(self.error)
 
+    def __deepcopy__(self, memo):
+        cls = self.__class__
+        result = cls.__new__(cls)
+        memo[id(self)] = result
+        for k, v in self.__dict__.items():
+            setattr(result, k, deepcopy(v, memo))
+        return result
```

Comment:
Maybe we're making things too complicated, what if we just do
```
        internal_dict = {k: v for k, v in self.items()}
        copied_dict = deepcopy(internal_dict, memo)
        result = self.__class__(copied_dict)
        memo[id(self)] = result
        return result
```
- i.e. rely on the dict's `deepcopy` mechanism and simply reconstruct a new Frozen object once that's done?

Reply:
In my understanding `memo` has to be fully configured further `deepcopy` calls to take advantage of the mechanism to avoid recursive calls, which would make it useless here. It's optional however, so it might be ok (at least for our use case) to just drop it.

Reply:
Right, but then in the original implementation of `__deepcopy__` in this PR, we don't actually check whether that ID is in `memo`, so then how is it supposed to prevent recursive calls? 🤔

Reply:
`deepcopy` does that, if `memo` is configured and provided, like so:
```python
        memo[id(self)] = result
        for k, v in self.__dict__.items():
            setattr(result, k, deepcopy(v, memo))
```

Reply:
I now have more questions, sorry 🥲  Why does `setattr` work at all, if our dict is supposed to be frozen?
And so basically it's fair to say that our frozen state is in conflict with the way `memo` should be used, no? If we need to make an "empty" object first, then put it in `memo`, then update it?

Reply:
> I now have more questions, sorry 🥲 Why does setattr work at all, if our dict is supposed to be frozen?

I'd say because `setattr()` adds a new attribute (overwriting an old one) and thus doesn't call `__setitem__` on an existing attribute in the dict.

> And so basically it's fair to say that our frozen state is in conflict with the way memo should be used, no? If we need to make an "empty" object first, then put it in memo, then update it?

How I understood it, the object should be created using `__new__`, not using `__init__` (hence my comment about the list `__deepcopy__()` not being quite right). So I think this should be correct for `FrozenDict`, but not for `FrozenList`. 
The practical implications are a different question - I guess we have to decide whether we may run into problems doing this.

Reply:
Ok, so honestly I don't think `setattr()` should be working at all, but that's a discussion for another time.

Thinking this through - I don't think reasonably there's any way you can get into recursiveness with frozen lists or dictionaries, because you have to define all values upon construction, at which time you don't have a reference to the final object yet, meaning also that you can't use it as value within itself.

Reply:
I do agree with that assessment. We could just get rid of using `memo` then?

Reply:
Yea, I think so. So just
```
internal_dict = {k: v for k, v in self.items()}
copied_dict = deepcopy(internal_dict, memo)
return self.__class__(copied_dict)
```
?

Reply:
I'd drop `memo` completely?
```suggestion
    def __deepcopy__(self, memo):
        internal_dict = {k: v for k, v in self.items()}
        return self.__class__(deepcopy(internal_dict))
```

Reply:
I'm not sure it matters, but I don't see the harm in passing the existing one onwards? 🤔

Reply:
You mean not doing a `deepcopy()` call? Not sure I follow.

Reply:
No I meant passing on the `memo` object to the `deepcopy()` call?